File: app/models/countdown.py
from datetime import datetime
from typing import Optional

class Countdown:
    """倒计时数据模型"""
    
    def __init__(
        self,
        id: Optional[int] = None,
        label: str = "",
        target_time: datetime = None,
        status: str = "running",
        remaining_seconds: int = 0,
        created_at: Optional[str] = None
    ):
        self.id = id
        self.label = label
        self.target_time = target_time or datetime.now()
        self.status = status
        self.remaining_seconds = remaining_seconds
        self.created_at = created_at or datetime.now().isoformat()
    
    def update_remaining(self):
        """更新剩余时间"""
        if self.status == "running":
            delta = self.target_time - datetime.now()
            self.remaining_seconds = max(0, int(delta.total_seconds()))
            if self.remaining_seconds == 0:
                self.status = "completed"
    
    def pause(self):
        """暂停倒计时"""
        if self.status == "running":
            self.update_remaining()
            self.status = "paused"
    
    def resume(self):
        """继续倒计时"""
        if self.status == "paused":
            self.target_time = datetime.now()
            self.status = "running"
    
    def reset(self, seconds: int):
        """重置倒计时"""
        self.remaining_seconds = seconds
        self.target_time = datetime.now()
        self.status = "running"
    
    def get_formatted_time(self) -> str:
        """获取格式化时间"""
        if self.status == "paused":
            seconds = self.remaining_seconds
        else:
            seconds = max(0, int((self.target_time - datetime.now()).total_seconds()))
        
        hours = seconds // 3600
        minutes = (seconds % 3600) // 60
        secs = seconds % 60
        
        if hours > 0:
            return f"{hours:02d}:{minutes:02d}:{secs:02d}"
        else:
            return f"{minutes:02d}:{secs:02d}"
    
    def to_dict(self) -> dict:
        """转换为字典"""
        return {
            'id': self.id,
            'label': self.label,
            'target_time': self.target_time.isoformat() if isinstance(self.target_time, datetime) else self.target_time,
            'status': self.status,
            'remaining_seconds': self.remaining_seconds,
            'created_at': self.created_at
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> 'Countdown':
        """从字典创建"""
        target_time = data.get('target_time')
        if isinstance(target_time, str):
            target_time = datetime.fromisoformat(target_time)
        
        return cls(
            id=data.get('id'),
            label=data.get('label', ''),
            target_time=target_time,
            status=data.get('status', 'running'),
            remaining_seconds=data.get('remaining_seconds', 0),
            created_at=data.get('created_at')
        )
    
    def __repr__(self):
        return f"Countdown(id={self.id}, label={self.label}, remaining={self.remaining_seconds}s, status={self.status})"

File: android/app/src/main/java/com/clock3/pet/data/model/Countdown.kt
package com.clock3.pet.data.model

import com.clock3.pet.data.entity.CountdownEntity
import com.clock3.pet.utils.AppLog
import java.time.Duration
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

data class Countdown(
    val id: Long = 0,
    val label: String,
    val targetTime: LocalDateTime,
    val status: CountdownStatus = CountdownStatus.RUNNING,
    var remainingSeconds: Int = 0,
    val createdAt: String? = null
) {
    enum class CountdownStatus {
        RUNNING, PAUSED, COMPLETED
    }

    fun updateRemaining() {
        if (status == CountdownStatus.RUNNING) {
            remainingSeconds = Duration.between(LocalDateTime.now(), targetTime).seconds.coerceIn(0, Int.MAX_VALUE.toLong()).toInt()
        }
    }

    fun getFormattedTime(): String {
        updateRemaining()
        val totalSeconds = remainingSeconds.coerceAtLeast(0)
        val hours = totalSeconds / 3600
        val minutes = (totalSeconds % 3600) / 60
        val seconds = totalSeconds % 60

        return if (hours > 0) {
            String.format("%02d:%02d:%02d", hours, minutes, seconds)
        } else {
            String.format("%02d:%02d", minutes, seconds)
        }
    }

    fun pause(): Countdown {
        if (status != CountdownStatus.RUNNING) return this
        val remaining = Duration.between(LocalDateTime.now(), targetTime).seconds.coerceIn(0, Int.MAX_VALUE.toLong()).toInt()
        return this.copy(
            status = CountdownStatus.PAUSED,
            remainingSeconds = remaining
        )
    }

    fun resume(): Countdown {
        if (status != CountdownStatus.PAUSED) return this
        return this.copy(
            targetTime = LocalDateTime.now().plusSeconds(remainingSeconds.toLong()),
            status = CountdownStatus.RUNNING
        )
    }

    fun toEntity(): CountdownEntity {
        return CountdownEntity(
            id = id,
            label = label,
            targetTime = targetTime.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME),
            status = status.name.lowercase(),
            remainingSeconds = remainingSeconds,
            createdAt = createdAt
        )
    }

    fun toExportMap(): Map<String, Any> {
        return mapOf(
            "id" to id,
            "label" to label,
            "targetTime" to targetTime.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME),
            "status" to status.name,
            "remainingSeconds" to remainingSeconds,
            "createdAt" to (createdAt ?: "")
        )
    }

    companion object {
        fun fromEntity(entity: CountdownEntity): Countdown {
            val status = try {
                Countdown.CountdownStatus.valueOf(entity.status.uppercase())
            } catch (e: Exception) {
                AppLog.w("Countdown", "Invalid status from DB: ${entity.status}")
                Countdown.CountdownStatus.COMPLETED
            }

            return Countdown(
                id = entity.id,
                label = entity.label,
                targetTime = LocalDateTime.parse(entity.targetTime, DateTimeFormatter.ISO_LOCAL_DATE_TIME),
                status = status,
                remainingSeconds = entity.remainingSeconds,
                createdAt = entity.createdAt
            )
        }

        fun fromExportMap(map: Map<*, *>): Countdown {
            val targetTimeStr = (map["targetTime"] as? String) ?: LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)
            val targetTime = try {
                LocalDateTime.parse(targetTimeStr, DateTimeFormatter.ISO_LOCAL_DATE_TIME)
            } catch (e: Exception) {
                LocalDateTime.now()
            }
            
            val rawStatus = (map["status"] as? String)?.let {
                try {
                    Countdown.CountdownStatus.valueOf(it.uppercase())
                } catch (e: Exception) {
                    Countdown.CountdownStatus.RUNNING
                }
            } ?: Countdown.CountdownStatus.RUNNING
            val finalStatus = if (rawStatus == CountdownStatus.RUNNING && targetTime.isBefore(LocalDateTime.now())) {
                CountdownStatus.COMPLETED
            } else {
                rawStatus
            }

            return Countdown(
                id = 0,
                label = (map["label"] as? String) ?: "倒计时",
                targetTime = targetTime,
                status = finalStatus,
                remainingSeconds = (map["remainingSeconds"] as? Number)?.toInt() ?: 0,
                createdAt = (map["createdAt"] as? String)?.let { if (it.isNotEmpty()) it else null }
            )
        }

        fun create(label: String, seconds: Int): Countdown {
            val safeSeconds = seconds.coerceAtLeast(0)
            return Countdown(
                label = label,
                targetTime = LocalDateTime.now().plusSeconds(safeSeconds.toLong()),
                remainingSeconds = safeSeconds
            )
        }
    }
}
